### PR TITLE
Split out arrow-cast (#2594)

### PR DESCRIPTION
**Pull request by tustvold**
# Which issue does this PR close?

Part of #2594

# Rationale for this change
 

The cast kernel is a fairly substantial piece of code, that takes a non-trivial time to compile. 

This is the last remaining kernel that the IO functionality, i.e. IPC, parquet, etc... depend on, and so this is a precursor to splitting them out.

# What changes are included in this PR?

Splits out the casting logic into a separate crate

# Are there any user-facing changes?

No

### CODE REVIEW

File: arrow-cast/src/cast.rs [hunk 1]
```diff
@@ -302,30 +293,31 @@ pub fn can_cast_types(from_type: &DataType, to_type: &DataType) -> bool {
 /// * To or from `StructArray`
 /// * List to primitive
 /// * Interval and duration
-pub fn cast(array: &ArrayRef, to_type: &DataType) -> Result<ArrayRef> {
+pub fn cast(array: &ArrayRef, to_type: &DataType) -> Result<ArrayRef, ArrowError> {
     cast_with_options(array, to_type, &DEFAULT_CAST_OPTIONS)
 }
 
-fn cast_integer_to_decimal128<T: ArrowNumericType>(
+fn cast_integer_to_decimal128<T: ArrowPrimitiveType>(
```
Comment (alamb) [inline arrow-cast/src/cast.rs:300]: is there a reason to also change this from `ArrowNumericType` to `ArrowPrimitiveType`?

Seems fine  as all ArrowNumericTypes` are also `ArrowPrimtiveTypes` https://docs.rs/arrow/latest/arrow/datatypes/trait.ArrowNumericType.html

File: arrow-cast/src/display.rs [hunk 2]
```diff
@@ -533,13 +529,23 @@ fn dict_array_value_to_string<K: ArrowPrimitiveType>(
...
-            row
-        ))
-    })?;
-
+    let dict_index = keys_array.value(row).as_usize();
     array_value_to_string(dict_array.values(), dict_index)
 }
+
+/// Converts numeric type to a `String`
+pub fn lexical_to_string<N: lexical_core::ToLexical>(n: N) -> String {
```
Comment (alamb) [inline arrow-cast/src/display.rs:537]: was just moved

**Review by alamb — APPROVED** [review]
LGTM -- this crate is going to be compiling full on parallelized before we know it

File: arrow-cast/src/cast.rs (hunk 1 above)
Comment (tustvold) [reply arrow-cast/src/cast.rs:300]: ArrowNumericType is part of the arithmetic kernels which haven't been split out

**Comment by tustvold** [comment]
Conflicts require #3018

**Comment by tustvold** [comment]
arrow-flight failure related to #3019

**Pull request merged**